File: 1218-lowest-common-ancestor-of-deepest-leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    pair<int, TreeNode*> dfs(TreeNode* node) {
        if (!node) return {0, nullptr};
        auto [l_h, l] = dfs(node->left);
        auto [r_h, r] = dfs(node->right);
        if (l_h > r_h) {
            return {l_h + 1, l};
        }
        else if (r_h > l_h) {
            return {r_h + 1, r};
        } else {
            return {r_h + 1, node};
        }
    }

    TreeNode* lcaDeepestLeaves(TreeNode* root) {
        return dfs(root).second;
    }
};

File: python/1218-lowest-common-ancestor-of-deepest-leaves.py
class Solution(object):
    def lcaDeepestLeaves(self, root):
        node, _ = self.dfs(root)
        return node
    def dfs(self, node):
        if not node:
            return None, 0
        left_node, left_height = self.dfs(node.left)
        right_node, right_height = self.dfs(node.right)
        if left_height > right_height:
            return left_node, left_height + 1
        elif right_height > left_height:
            return right_node, right_height + 1
        else:
            return node, left_height + 1